modules/currency.js

```diff
@@ -0,0 +1,206 @@
+
+import bidfactory from 'src/bidfactory';
+import { STATUS } from 'src/constants';
+import { ajax } from 'src/ajax';
+import * as utils from 'src/utils';
+import bidmanager from 'src/bidmanager';
+
+$$PREBID_GLOBAL$$.currency = setConfig;
+
+const DEFAULT_CURRENCY_RATE_URL = "http://prebid.aws.rubiconproject.com/latest.json";
+const CURRENCY_RATE_PRECISION = 4;
+
+var bidResponseQueue = [];
+var conversionCache = {};
+var currencyRatesLoaded = false;
+var adServerCurrency = 'USD';
+
+var oldBidResponse;
+
+export var currencySupportEnabled = false;
+export var currencyRates = {};
+
+export function setConfig(config) {
+  var url = DEFAULT_CURRENCY_RATE_URL;
+  if (typeof config.adServerCurrency === 'string') {
+    utils.logInfo('enabling currency support', arguments);
+
+    adServerCurrency = config.adServerCurrency;
+    if (config.conversionRateFile) {
+      utils.logInfo('currency using override conversionRateFile:', config.conversionRateFile);
+      url = config.conversionRateFile;
+    }
+    initCurrency(url);
+  } else {
+    // currency support is disabled, setting defaults
+    utils.logInfo('disabling currency support');
+    resetCurrency();
+  }
+}
+
+function initCurrency(url) {
+  utils.logInfo('Invoking initCurrency', arguments);
+
+  conversionCache = {};
+  currencySupportEnabled = true;
+
+  if(!oldBidResponse) {
+    utils.logInfo('Installing addBidResponse decorator for currency module', arguments);
+
+    oldBidResponse = bidmanager.addBidResponse;
+    bidmanager.addBidResponse = addBidResponseDecorator(bidmanager.addBidResponse);
+  }
+
+  ajax(url, function(response) {
+    try {
+      currencyRates = JSON.parse(response);
+      utils.logInfo('currencyRates set to ' + JSON.stringify(currencyRates));
+    } catch (e) {
+      utils.logError('failed to parse currencyRates response: ' + response);
+    }
+    currencyRatesLoaded = true;
+    processBidResponseQueue();
+  });
+}
+
+function resetCurrency() {
+  if(oldBidResponse) {
+    utils.logInfo('Uninstalling addBidResponse decorator for currency module', arguments);
+
+    bidmanager.addBidResponse = oldBidResponse;
+    oldBidResponse = undefined;
+  }
+
+  adServerCurrency = 'USD';
+  conversionCache = {};
+  currencySupportEnabled = false;
+  currencyRatesLoaded = false;
+  currencyRates = {};
+}
+
+export function addBidResponseDecorator(fn) {
+  return function(adUnitCode, bid) {
+    utils.logInfo('Invoking addBidResponseDecorator function', arguments);
+
+    // default to USD if currency not set
+    if(bid && !bid.currency) {
```

Comment:
can add checks here and line 91 at the top ( `if(!bid)` ) and exit immediately to reduce code complexity.